samza-azure/src/main/java/org/apache/samza/system/azureblob/producer/AzureBlobSystemProducer.java

```diff
@@ -0,0 +1,505 @@
+    Preconditions.checkNotNull(systemName, "System name can not be null when creating AzureBlobSystemProducer");
+    Preconditions.checkNotNull(config, "Config can not be null when creating AzureBlobSystemProducer");
+    Preconditions.checkNotNull(metricsRegistry, "Metrics registry can not be null when creating AzureBlobSystemProducer");
+
+    // Azure logs do not show without this property set
+    System.setProperty(Configuration.PROPERTY_AZURE_LOG_LEVEL, "1");
+    this.systemName = systemName;
+    this.config = config;
+
+    String writerFactoryClassName = this.config.getAzureBlobWriterFactoryClassName(this.systemName);
+    try {
+      this.writerFactory = (AzureBlobWriterFactory) Class.forName(writerFactoryClassName).newInstance();
+    } catch (Exception e) {
+      throw new SystemProducerException("Could not create writer factory with name " + writerFactoryClassName, e);
+    }
+    this.flushTimeoutMs = this.config.getFlushTimeoutMs(this.systemName);
+    this.closeTimeout = this.config.getCloseTimeoutMs(this.systemName);
+    this.blockFlushThresholdSize = this.config.getMaxFlushThresholdSize(this.systemName);
+    int asyncBlobThreadPoolCount = this.config.getAzureBlobThreadPoolCount(this.systemName);
+    int blockingQueueSize = this.config.getBlockingQueueSizeOrDefault(this.systemName, asyncBlobThreadPoolCount * 2);
+
+    LOG.info("SystemName: {} block flush size:{}", systemName, this.blockFlushThresholdSize);
+    LOG.info("SystemName: {} thread count:{}", systemName, asyncBlobThreadPoolCount);
+
+    BlockingQueue<Runnable>
+        linkedBlockingDeque = new LinkedBlockingDeque<>(blockingQueueSize);
+
+    this.asyncBlobThreadPool =
+        new ThreadPoolExecutor(asyncBlobThreadPoolCount, asyncBlobThreadPoolCount, 60,
+            TimeUnit.SECONDS, linkedBlockingDeque, new ThreadPoolExecutor.CallerRunsPolicy());
+
+    this.writerMap = new ConcurrentHashMap<>();
+
+    this.metrics = new AzureBlobSystemProducerMetrics(systemName, config.getAzureAccountName(systemName), metricsRegistry);
+  }
+
+  /**
+   * {@inheritDoc}
+   * @throws SystemProducerException
+   */
+  @Override
+  public synchronized void start() {
+    if (isStarted) {
+      throw new SystemProducerException("Attempting to start an already started producer.");
+    }
+
+    String accountName = config.getAzureAccountName(systemName);
+    String accountKey = config.getAzureAccountKey(systemName);
+
+    setupAzureContainer(accountName, accountKey);
+
+    LOG.info("Starting producer.");
+    isStarted = true;
+  }
+
+  /**
+   * {@inheritDoc}
+   * @throws SystemProducerException
+   */
+  @Override
+  public synchronized void stop() {
+    if (!isStarted) {
+      throw new SystemProducerException("Attempting to stop a producer that was not started.");
+    }
+
+    if (isStopped) {
+      throw new SystemProducerException("Attempting to stop an already stopped producer.");
+    }
```

Comment:
Should this throw an exception in these cases? What do other `SystemProducer` impls do?

Reply:
My question is if you should throw an exception or just ignore the call. The interface docs don't seem to indicate if a failure should actually be thrown or not when calling `stop` twice.